### PR TITLE
Shared context with plugin apps

### LINKED ISSUE #2148: Plugins: shared context
## Is your feature request related to a problem? Please describe 👀

With the implementation of the webpack module federation, the react contexts are not shared to the plugins.
This means that the current language, theme etc are unavailable to plugins.

## Describe the solution you'd like 🎁
The context should be shared to all plugins
Areas are:

 - [x] Material UI theme Provider
 - [x] Intl Provider
 - [x] QueryClientProvider
 - [x] AuthProvider
 - [x] AlertModalProvider
 - [x] ConfirmationModalProvider
 - [ ] Router

### Moneyworks Jobcode 🧰

`OMS:PLUG`

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2148

# 👩🏻‍💻 What does this PR do? 
 
The standard Provider / Context model is not working across the plugin apps because they are loaded dynamically.
We need to be able to share a number of aspects though, so I've implemented `react-singleton-context` in order to share contexts.

Created provider proxies for these, as the context is handled internally within the components and I'm unable to replace the `useContext` call. The end result is that the plugin must wrap components in a proxy component, which fetches the shared context and populates the provider which is then used by the child components.
 - [ ] Material UI theme Provider
 - [ ] QueryClientProvider

These ones were simpler and work without anything extra in the plugin:
 - [ ] Intl Provider
 - [ ] AuthProvider
 - [ ] AlertModalProvider
 - [ ] ConfirmationModalProvider

And the `BrowserRouter` I gave up on. I was unable to share the router / routes and so have just added a `BrowserRouter` into the plugin, so that any child calls to `useNavigate()` etc won't fail. While this prevents any errors, the browser is not actually navigated because the BrowserRouter instance used by the plugin doesn't have the routes populated. Possibly a task for further exploration.

# 🧪 How has/should this change been tested? 

I've used the [dashboard plugins example](https://github.com/openmsupply/open-msupply/tree/feature/front-end-plugins-example-01) to test out the various features of this one. 

The dashboard plugins are making use of theme, query client, intl, auth and router. I had added a quick AlertModal and Confirmation modal to test those as well.

To test without building the plugins, you can run `yarn build` from the root folder to build both client and server. Then copy the contents of the plugins into `server/app_data/plugins` and see what happens

[ShippingStatus.zip](https://github.com/openmsupply/open-msupply/files/12482571/ShippingStatus.zip)
[Dashboard.zip](https://github.com/openmsupply/open-msupply/files/12482569/Dashboard.zip)

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: client/packages/common/src/intl/utils/IntlUtils.ts [hunk 1]
```diff
@@ -49,12 +50,12 @@ export const useIntlUtils = () => {
     }
 
     // Handle languages such as en-US or fr-FR
-    const baseLanguage = supportedLanguage.split('-')[0] as SupportedLocales;
+    const baseLanguage = supportedLanguage?.split('-')[0] as SupportedLocales;
```
Comment (mark-prins) [inline client/packages/common/src/intl/utils/IntlUtils.ts:53]: this and the one below are partly because the context is initially empty and this will cause a problem on the first render.

File: client/packages/common/src/plugins/types.ts [hunk 2]
```diff
@@ -1,11 +1,13 @@
 export enum PluginArea {
   AppBar,
+  Dashboard,
   Toolbar,
 }
 
 export enum PluginType {
   InboundShipment,
   InternalOrder,
+  None,
```
Comment (mark-prins) [inline client/packages/common/src/plugins/types.ts:10]: Added this for the plugins which don't require data to be passed in

File: client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx [hunk 3]
```diff
@@ -50,7 +55,9 @@ export const BaseDatePickerInput: FC<
   const t = useTranslation('common');
 
   const dateValue =
-    typeof props.value === 'string' ? new Date(props.value) : props.value;
+    typeof props.value === 'string'
```
Comment (mark-prins) [inline client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:58]: oops. this is an offroad! I fixed #2153 because it was causing an issue while testing a different issue.

File: client/packages/common/src/intl/strings/IntlStrings.ts [hunk 4]
```diff
@@ -18,7 +19,8 @@ export interface TypedTFunction<Keys> {
 //
 // returned function can be used with optional ns, i.e. t('label.create-user', { ns: 'system' })
 export const useTranslation = (ns?: Namespace): TypedTFunction<LocaleKey> => {
-  const { t } = useTranslationNext(ns);
+  const {i18n} = useIntl();
+  const { t } = useTranslationNext(ns, {i18n});
```
Comment (clemens-msupply) [inline client/packages/common/src/intl/strings/IntlStrings.ts:23]: Why was this changed?

File: client/packages/common/src/plugins/types.ts (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/common/src/plugins/types.ts:10]: This is not clear to me, do you have an example?

**Review by clemens-msupply — COMMENTED** [review]
~Could you please explain this a bit more?
`These ones were simpler and work without anything extra in the plugin:`
Why are they simpler? what is the difference?~

Edit: think I got it

I tried to copied the provided ShippingStatus plugin to the server and just started the FE with `yarn start-local` but for some reason the plugin doesn't get loaded. The api returns the plugin though...

File: client/packages/common/src/plugins/types.ts (hunk 2 above)
Comment (mark-prins) [reply client/packages/common/src/plugins/types.ts:10]: The [dashboard plugins](https://github.com/openmsupply/open-msupply/tree/feature/front-end-plugins-example-01) are making use of this - I had assumed that all plugins would require context-related data.. but the dashboard is an exception. The components are fetching stats independently

File: client/packages/common/src/intl/strings/IntlStrings.ts (hunk 4 above)
Comment (mark-prins) [reply client/packages/common/src/intl/strings/IntlStrings.ts:23]: in order to share the i18n object between the host app and the plugins - to do so, I've had to keep the `i18n` in a shared context, and here retrieve it and us that as an option when calling `useTranslation`. The original problem is that the host app is able to translate, and knows the current language etc - but the plugins are operating in a different context and they are unable to see the current intl configuration.

File: client/packages/common/src/plugins/types.ts (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/common/src/plugins/types.ts:10]: Maybe you could call it Dashboard? and in PluginArea you call it DashboardWidget or similar?

Comment (mark-prins) [reply client/packages/common/src/plugins/types.ts:10]: I do like that - but it does mean that `Dashboard` implies no data. I think it is better to explicitly state `No data` in some way. `Fetches it's own data`? 
Oh wait.. perhaps the issue is that I've called it `PluginType` when I had in my mind `PluginDataType`. I'll change it to `Dashboard` 🙏

File: client/packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx [hunk 5]
```diff
@@ -31,6 +34,11 @@ export const AppBarButtonsComponent: FC<AppBarButtonProps> = ({
   const { OpenButton } = useDetailPanel();
   const t = useTranslation('common');
   const { print, isPrinting } = useReport.utils.print();
+  const pluginButtons = usePluginElements({
+    area: PluginArea.AppBar,
```
Comment (clemens-msupply) [inline client/packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx:38]: I believe this is Toolbar? The provided plugin then loads but crashes... Not sure if I have an outdated plugin though...

File: client/packages/host/src/PluginsProvider.tsx [hunk 6]
```diff
@@ -19,8 +19,19 @@ const getPluginArea = (area: string) => {
 
 const getPluginType = (type: string) => {
   switch (type) {
+    case 'dashboard':
```
Comment (clemens-msupply) [inline client/packages/host/src/PluginsProvider.tsx:22]:
```suggestion
    case 'Dashboard':
```

This wasn't working for me...

**Review by clemens-msupply — APPROVED** [review]
Looks good as far as I was able to test. Thanks!

Had to fix a typo in PluginProvider to load the dashboard plugin (see comment)
Think `getPluginArea` need to be updated as well?

File: client/packages/host/src/PluginsProvider.tsx (hunk 6 above)
Comment (mark-prins) [reply client/packages/host/src/PluginsProvider.tsx:22]: yes! sorry - had a note to update that and haven't done it yet

File: client/packages/common/src/api/QueryClientProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  PropsWithChildren<{ client: QueryClient }>
+> = ({ client, children }) => {
+  return (
+    <QueryClientContext.Provider value={{ client }}>
+      <QueryClientProviderProxy>{children}</QueryClientProviderProxy>
+    </QueryClientContext.Provider>
+  );
+};
+
+export const QueryClientProviderProxy: FC<PropsWithChildrenOnly> = ({
```
Comment (andreievg) [inline client/packages/common/src/api/QueryClientProvider.tsx:24]: Was wondering, is this just a stylistic preference or am i missing something, vs:

```react
export const QueryClientProvider: FC<
  PropsWithChildren<{ client: QueryClient }>
> = ({ client, children }) => {
  return (
    <QueryClientContext.Provider value={{ client }}>
      <ReactQueryClientProvider client={client}>
         {children}
      </ReactQueryClientProvider>
    </QueryClientContext.Provider>
  );
};

```

**Review by andreievg — COMMENTED** [review]
Looks good, I was wondering about the need of proxy context. Was also wondering how we do module federation with zustand (I've started to like it and personally would have used zustand instead of contexts, were possible)

**Comment by mark-prins** [comment]
> I was wondering about the need of proxy context

this is unfortunately required because of components which are outside of our control - such as the MUI components which are looking for a theme provider / context somewhere above them in the hierarchy. We have that configured in the host app - the plugins don't see that context though, so I've had to create the proxy which shares across the apps

really like the idea of using zustand to have a shared state. will need to run some tests and see if that works as expected

File: client/packages/common/src/api/QueryClientProvider.tsx (hunk 7 above)
Comment (mark-prins) [reply client/packages/common/src/api/QueryClientProvider.tsx:24]: no, not a stylistic preference. the plugin needs to implement the Context, while also having access to the `client` which is initialised in the host app.

To do that, I've had to use the `react-singleton-context` which stores the `client`. However, existing components (from the packages, such as QueryClient) are looking for their own Context/Provider - and will not 'see' the custom one. So, I've had to store the state in the higher level, then fetch the state from that higher level in the Proxy which then instantiates the type of Provider that the package is trying to find.
